lib/bitcoin/versions/latest/NormalizedFeeCalculator.ts

```diff
@@ -8,55 +8,76 @@ import IFeeCalculator from '../../interfaces/IFeeCalculator';
     console.log(`Initializing normalized fee calculator.`);
   }
 
-  /**
-   * This adds normalized fee to the block as it would calculate normalziedFee, but uses a cache to remeber previously seen blocks.
-   * Which reduces calls to the metadata store.
-   */
   public async addNormalizedFeeToBlockMetadata (blockMetadata: BlockMetadataWithoutNormalizedFee): Promise<BlockMetadata> {
+
+    // If the height of the given block does not have large enough look-back window, just use initial fee.
     if (blockMetadata.height < this.genesisBlockNumber + this.feeLookBackWindowInBlocks) {
-      return Object.assign({ normalizedFee: this.initialNormalizedFee }, blockMetadata);
+      const blockWithFee = Object.assign({ normalizedFee: this.initialNormalizedFeeInSatoshis }, blockMetadata);
+
+      // We need to push the block metadata into the look-back cache in preparation for when look-back window becomes large enough with the given block height.
+      this.cachedLookBackWindow.push(blockWithFee);
+      this.blockHeightOfCachedLookBackWindow = blockMetadata.height + 1;
+
+      return blockWithFee;
     }
-    // the cache won't work if the block is not expected, refetch the blocks and store in cache
-    if (this.expectedBlockHeight !== blockMetadata.height) {
-      this.blockMetadataCache = await this.getBlocksInLookBackWindow(blockMetadata.height);
-      this.expectedBlockHeight = blockMetadata.height;
+
+    // Code reaches here whn the look-back window is large enough.
+
+    // The cache won't work if the block is not the anticipated height, refetch the blocks and store in cache.
+    if (this.blockHeightOfCachedLookBackWindow !== blockMetadata.height) {
+      this.cachedLookBackWindow = await this.getBlocksInLookBackWindow(blockMetadata.height);
+      this.blockHeightOfCachedLookBackWindow = blockMetadata.height;
     }
-    const newFee = this.calculateNormalizedFee(this.blockMetadataCache);
-    const newBlockWithFee = Object.assign({ normalizedFee: newFee }, blockMetadata);
-    this.blockMetadataCache.push(newBlockWithFee);
-    this.blockMetadataCache.shift();
-    this.expectedBlockHeight++;
+
+    const normalizedFee = this.calculateNormalizedFee(this.cachedLookBackWindow);
+    const newBlockWithFee = Object.assign({ normalizedFee }, blockMetadata);
+    console.log(normalizedFee);
+    this.cachedLookBackWindow.push(newBlockWithFee);
+    this.cachedLookBackWindow.shift();
+    this.blockHeightOfCachedLookBackWindow++;
     return newBlockWithFee;
   }
 
   public async getNormalizedFee (block: number): Promise<number> {
     if (block < this.genesisBlockNumber + this.feeLookBackWindowInBlocks) {
-      return this.initialNormalizedFee;
+      return this.initialNormalizedFeeInSatoshis;
     }
     const blocksToAverage = await this.getBlocksInLookBackWindow(block);
-    return this.calculateNormalizedFee(blocksToAverage);
+    
+    // TODO: @Isaac, why are we not getting the content fee from DB?
```

Comment:
We can. I don't think it will happen, but I was thinking if core asks for 1 block ahead, then we will need to calculate it, or else it will be undefined. Probably will never happen.

Reply:
Will change in my pr

Reply:
Synced offline.